executor/aggfuncs/aggfunc_test.go

```diff
@@ -975,37 +783,96 @@ func testAggMemFunc(t *testing.T, p aggMemTest) {
 	}
 }
 
-// Deprecated: migrating to testAggMemFunc(t *testing.T, p aggMemTest)
-func (s *testSuite) testAggMemFunc(c *C, p aggMemTest) {
-	srcChk := p.aggTest.genSrcChk()
+func testMultiArgsAggFunc(t *testing.T, ctx sessionctx.Context, p multiArgsAggTest) {
+	srcChk := p.genSrcChk()
 
-	args := []expression.Expression{&expression.Column{RetType: p.aggTest.dataType, Index: 0}}
-	if p.aggTest.funcName == ast.AggFuncGroupConcat {
+	args := make([]expression.Expression, len(p.dataTypes))
+	for k := 0; k < len(p.dataTypes); k++ {
+		args[k] = &expression.Column{RetType: p.dataTypes[k], Index: k}
+	}
+	if p.funcName == ast.AggFuncGroupConcat {
 		args = append(args, &expression.Constant{Value: types.NewStringDatum(separator), RetType: types.NewFieldType(mysql.TypeString)})
 	}
-	desc, err := aggregation.NewAggFuncDesc(s.ctx, p.aggTest.funcName, args, p.isDistinct)
-	c.Assert(err, IsNil)
-	if p.aggTest.orderBy {
+
+	desc, err := aggregation.NewAggFuncDesc(ctx, p.funcName, args, false)
+	require.NoError(t, err)
+	if p.orderBy {
 		desc.OrderByItems = []*util.ByItems{
 			{Expr: args[0], Desc: true},
 		}
 	}
-	finalFunc := aggfuncs.Build(s.ctx, desc, 0)
-	finalPr, memDelta := finalFunc.AllocPartialResult()
-	c.Assert(memDelta, Equals, p.allocMemDelta)
+	finalFunc := aggfuncs.Build(ctx, desc, 0)
+	finalPr, _ := finalFunc.AllocPartialResult()
+	resultChk := chunk.NewChunkWithCapacity([]*types.FieldType{desc.RetTp}, 1)
 
-	updateMemDeltas, err := p.updateMemDeltaGens(srcChk, p.aggTest.dataType)
-	c.Assert(err, IsNil)
 	iter := chunk.NewIterator4Chunk(srcChk)
-	i := 0
 	for row := iter.Begin(); row != iter.End(); row = iter.Next() {
-		memDelta, err := finalFunc.UpdatePartialResult(s.ctx, []chunk.Row{row}, finalPr)
-		c.Assert(err, IsNil)
-		c.Assert(memDelta, Equals, updateMemDeltas[i])
-		i++
+		// FIXME: cannot assert error since there are cases of error, e.g. rows were cut by GROUPCONCAT
+		_, _ = finalFunc.UpdatePartialResult(ctx, []chunk.Row{row}, finalPr)
```

Comment:
Is it a bug? Otherwise I think this PR is beyond a migration but you change the test logic nontrivially.

Reply:
Thank you for your quick review,

I have not changed the test logic. I think the results shown in the diff on this PR is misleading. The diff result **compares `testMultiArgsAggFunc` function and `(s *testSuite) testAggMemFunc` method**.

```diff
- // Deprecated: migrating to testAggMemFunc(t *testing.T, p aggMemTest)
- func (s *testSuite) testAggMemFunc(c *C, p aggMemTest) {
- // ....
- 	for row := iter.Begin(); row != iter.End(); row = iter.Next() {
-		memDelta, err := finalFunc.UpdatePartialResult(s.ctx, []chunk.Row{row}, finalPr)
-		c.Assert(err, IsNil)
-		c.Assert(memDelta, Equals, updateMemDeltas[i])
-		i++
- // ...
- }

+ // This logic same (s *testSuite) testMultiArgsAggMemFunc
+ func testMultiArgsAggFunc(t *testing.T, ctx sessionctx.Context, p multiArgsAggTest) {
+ // ...

+	for row := iter.Begin(); row != iter.End(); row = iter.Next() {
+		// FIXME: cannot assert error since there are cases of error, e.g. rows were cut by GROUPCONCAT
+		_, _ = finalFunc.UpdatePartialResult(ctx, []chunk.Row{row}, finalPr)
+ // ...
+ }

+ // Deprecated: migrating to testMultiArgsAggMemFunc(t *testing.T, p multiArgsAggMemTest)
  func (s *testSuite) testMultiArgsAggMemFunc(c *C, p multiArgsAggMemTest) {
	for row := iter.Begin(); row != iter.End(); row = iter.Next() {
		// FIXME: cannot assert error since there are cases of error, e.g. rows were cut by GROUPCONCAT
		_, _ = finalFunc.UpdatePartialResult(ctx, []chunk.Row{row}, finalPr)
```

## Details

- I have added `testMultiArgsAggFunc` function by clone `(s *testSuite) testMultiArgsAggFunc` method. Function `testMultiArgsAggFunc` did not change the test logic.
    - `(s *testSuite) testMultiArgsAggFunc` method could not be deleted because it was called from a test file.
    - ref: https://github.com/pingcap/tidb/pull/29529#discussion_r744197085
- Then I removed `(s *testSuite) testAggMemFunc` method, because no one calls it anymore.
    - https://github.com/pingcap/tidb/pull/29529/commits/acd7f053ae2f346871243623c597a07005fe2760#diff-7cf602fc7a7da0db3c65a5304060ad46df30b6e6f4c6a47b678e7b88bc2e7635L978-L1008

The diff result **compares `testMultiArgsAggFunc` function and `(s *testSuite) testAggMemFunc` method**, but The correct comparison is shown in the following screenshot.


- `func (s *testSuite) testMultiArgsAggFunc(c *C, p multiArgsAggTest)`
    - https://github.com/pingcap/tidb/blob/7ae8ce7b155cf06a2c03e23a1ff6771bd4e60a71/executor/aggfuncs/aggfunc_test.go#L898-L903
- `func testMultiArgsAggFunc(t *testing.T, ctx sessionctx.Context, p multiArgsAggTest)`
    - https://github.com/pingcap/tidb/blob/7ae8ce7b155cf06a2c03e23a1ff6771bd4e60a71/executor/aggfuncs/aggfunc_test.go#L808-L813

![image](https://user-images.githubusercontent.com/1883265/140670076-55208e8d-357d-4fb8-bb4d-12d0bc6f798b.png)

Reply:
Nice, I don't need to compare by my eyes.